elephant/asset.py

```diff
@@ -485,20 +568,13 @@ def _jsf_uniform_orderstat_3d(u, n, verbose=False):
     # using matrix algebra
     # initialise probabilities to 0
     P_total = np.zeros(du.shape[0], dtype=np.float32)
-    iter_id = 0
-    for matrix_entries in tqdm(itertools.product(*lists),
-                               total=it_todo,
-                               desc="Joint survival function",
-                               disable=not verbose):
+    for iter_id, matrix_entries in enumerate(
+            tqdm(_combinations_with_replacement(n, d=d),
+                 total=it_todo,
+                 desc="Joint survival function",
+                 disable=not verbose)):
         # if we are running with MPI
         if mpi_accelerated and iter_id % size != rank:
-            iter_id += 1
-            continue
-
-        iter_id += 1
-
-        # test for valid pyramid and exit loop early
-        if _wrong_order(matrix_entries):
             continue
 
         # we only need the differences of the indices:
```

Comment:
This statement also takes a lot of time during the loop.

```
import numpy as np
import timeit
from functools import partial

def test_old(n):
   matrix_entries = (3, 2, 1)
   di = -np.diff(matrix_entries, prepend=n, append=0)
   return di


def test_alt(n):
    matrix_entries = (3, 2, 1)
    di = -np.diff((n,) + matrix_entries + (0,))
    return di

n = 5
print(test_old(n), timeit.timeit(partial(test_old, n)))
print(test_alt(n), timeit.timeit(partial(test_alt, n)))
```

which produces

```
[2 1 1 1] 17.07817635399988
[2 1 1 1] 4.9544372519885655
```

Reply:
I can confirm a persistent speedup.

n=20
|d|_jsf_uniform_orderstat_3d|_jsf_uniform_orderstat_3d_manually_add|delta|speedup|
|---|---|---|---|---|
|2|0.03|0.02|0.01|1.27|
|3|0.18|0.14|0.04|1.28|
|4|1.16|0.90|0.26|1.28|
|5|12.51|10.19|2.31|1.23|
|6|53.32|44.03|9.29|1.21|
|7|188.27|154.97|33.30|1.21|
|8|750.50|614.45|136.05|1.22|
|9|2098.58|2017.96|80.62|1.04|
|10|5983.85|5022.03|961.82|1.19|
|11|16010.13|14015.33|1994.80|1.14|
|12|41783.34|34124.50|7658.84|1.22|
|13|93831.37|80599.86|13231.51|1.16|

n=200
|d|_jsf_uniform_orderstat_3d|_jsf_uniform_orderstat_3d_manually_add|delta|speedup|
|---|---|---|---|---|
|2|3.28|2.54|0.74|1.29|
|3|356.36|292.51|63.85|1.22|
|4|18767.29|14902.68|3864.61|1.26|